katdal/chunkstore.py

```diff
@@ -456,19 +470,62 @@ def get_dask_array(self, array_name, chunks, dtype, offset=(), errors=0):
             Dask array of given dtype
         """
         kwargs = {'dtype': dtype}
-        if errors == 'none':
-            get_func = self.get_chunk_or_none
+        if errors == 'placeholder':
+            get_func = self.get_chunk_or_placeholder
         elif errors == 'raise':
             get_func = self.get_chunk
+        elif isinstance(errors, str):
+            raise ValueError("Unexpected value for errors; expected 'placeholder', 'raise', or a number")
         else:
             get_func = self.get_chunk_or_default
             kwargs['default_value'] = errors
         getter = functools.partial(get_func, **kwargs)
-        if offset:
+
+        if index:
+            chunks = [list(c) for c in chunks]   # Make mutable
+            shape = [sum(c) for c in chunks]
+            index = list(da.slicing.normalize_index(index, shape))
+            if not all(isinstance(idx, slice) and idx.step is None
+                       for idx in index):
+                raise IndexError('Only slices with unit step are valid indices in get_dask_array')
+            if not offset:
+                offset = [0] * len(shape)
+            else:
+                offset = list(offset)
+            for axis in range(len(shape)):
+                if index[axis] == slice(None):
+                    continue
+                start, stop, step = index[axis].indices(shape[axis])
+                assert step == 1
+                # Remove unneeded chunks from the ends
+                start_chunk = 0
+                while start_chunk < len(chunks[axis]) and chunks[axis][start_chunk] <= start:
+                    c = chunks[axis][start_chunk]
+                    offset[axis] += c
+                    start -= c
+                    stop -= c
+                    shape[axis] -= c
+                    start_chunk += 1
+                stop_chunk = len(chunks[axis])
+                while stop_chunk > start_chunk and chunks[axis][stop_chunk - 1] <= shape[axis] - stop:
+                    stop_chunk -= 1
+                    c = chunks[axis][stop_chunk]
+                    shape[axis] -= c
+                chunks[axis] = chunks[axis][start_chunk : stop_chunk]
+                if not chunks[axis]:
+                    chunks[axis] = (0,)   # Dask doesn't allow empty chunk lists
+                index[axis] = slice(start, stop)
+            # Go back to tuples
+            chunks = tuple(tuple(c) for c in chunks)
+            index = tuple(index)
+            offset = tuple(offset)
+
+        if any(offset):
             getter = _add_offset_to_slices(getter, offset)
         # Use dask utility function that forms the core of da.from_array
         dask_graph = da.core.getem(array_name, chunks, getter)
-        return da.Array(dask_graph, array_name, chunks, dtype)
+        array = da.Array(dask_graph, array_name, chunks, dtype)
+        return array[index]
```

Comment:
So `index` will be trivial if the selection is aligned with the original chunks?

Reply:
I'm not entirely sure what you mean by this. But `Array.__getitem__` has a check for a no-op index and just returns the original array without creating new layers.

Reply:
When I initially looked at this, it gave the impression that you are generating a bigger graph and then just selecting a part like before. This is before I worked my way through the rechunking code :-)

I then wondered why you would still need `index` but eventually realised that selecting e.g. channel 1 to 13 will require fractional chunks and therefore further indexing. But with a chunk per dump and a range of selected channels commensurate with the chunking, this `index` will basically be `[:, :]`. That's the part I'm asking about :-)

Reply:
Right, I'm with you now. Yes, this is here in case for the case of fractional chunks.